Mapbox v10

grzesiek2010 opened the pull request:
Closes #4893 
~~Blocked by #5139~~

Most of the discussion around this PR is in the issue or here so no need to write this up really.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

grzesiek2010 committed:
Added mapbox v10 dependency

grzesiek2010 committed:
Removed mapbox v9 from local maven repo

grzesiek2010 committed:
Fixed imports in Defaults class

grzesiek2010 committed:
Fixed imports in MapProvider class

grzesiek2010 committed:
Fixed MapBoxInitializationFragment and converted it to kotlin

grzesiek2010 committed:
Reworked MapboxMapFragment

grzesiek2010 committed:
Removed MapboxUtils class

grzesiek2010 committed:
Removed excluding mapbox libs

grzesiek2010 committed:
Removed unused MapFragment

grzesiek2010 committed:
Converted MapboxLocationCallback to kotlin

grzesiek2010 committed:
Removed unused empty.xml file

grzesiek2010 committed:
Fixed build

grzesiek2010 commented:
Ok @lognaturel @seadowg it's time to decide how to handle Mapbox 10. As we talked recently options are:
- sharing our tokens so that everyone can build the app including CircleCi
- storing Mapbox 10 in our local repository
- having a separate flavor

I was more in favor of one of two first options because it would be easier. Then if it doesn't work for us or there is any problem with licenses we can add a separate flavor.
Please think about it one more time and we can discuss it.

grzesiek2010 commented on collect_app/build.gradle:
```diff
@@ -177,13 +173,6 @@ android {
-        // the Mapbox native library only for 32-bit and 64-bit ARM devices and
-        // omit it for all X86 devices.
-        exclude 'lib/x86/libmapbox-gl.so'
-        exclude 'lib/x86_64/libmapbox-gl.so'
```
I think you are not going to like it. Mapbox 10 has libs with different naming. I could of course update it here and delete too but that causes crashes when the app is started so it looks like we no longer can do tricks like that. That increases the size of our apk of course...

seadowg replied:
What crash are you seeing? I think we'll want to do as much as can to avoid including unused `x86` libs.

lognaturel replied:
Agreed with @seadowg that more details are needed here. What's the binary size without the filtering?

grzesiek2010 replied:
The error is:
```
Caused by: java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/org.odk.collect.android-fpP4T_tgz8x_GU7w9kKqvA==/base.apk"],nativeLibraryDirectories=[/data/app/org.odk.collect.android-fpP4T_tgz8x_GU7w9kKqvA==/lib/x86_64, /data/app/org.odk.collect.android-fpP4T_tgz8x_GU7w9kKqvA==/base.apk!/lib/x86_64, /system/lib64, /system/product/lib64]]] couldn't find "libmapbox-maps.so"
        at java.lang.Runtime.loadLibrary0(Runtime.java:1067)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1007)
        at java.lang.System.loadLibrary(System.java:1667)
        at com.mapbox.common.loader.MapboxLibraryLoader.load(MapboxLibraryLoader.kt:19)
        at com.mapbox.maps.loader.MapboxMapsInitializer.create(MapboxMapsInitializer.java:19)
        at com.mapbox.maps.loader.MapboxMapsInitializer.create(MapboxMapsInitializer.java:16)
        at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:180)
        at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:238) 
        at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:206) 
        at androidx.startup.InitializationProvider.onCreate(InitializationProvider.java:45) 
        at android.content.ContentProvider.attachInfo(ContentProvider.java:2092) 
        at android.content.ContentProvider.attachInfo(ContentProvider.java:2066) 
        at android.app.ActivityThread.installProvider(ActivityThread.java:6983) 
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6528) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6445) 
        at android.app.ActivityThread.access$1300(ActivityThread.java:219) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859) 
```

it takes place immediately when the app is started 

Here is the size of those files:
![image](https://user-images.githubusercontent.com/3276264/168651755-ff04ec9e-bb37-429c-8bc7-93e3069f9f06.png)

lognaturel replied:
If you haven't already, could you please review https://github.com/getodk/collect/pull/3108 and https://github.com/getodk/collect/pull/3118 to see if there are any ideas there?

The error you shared looks like it would be from a x86_64 device. Was it? I would not expect an arm device to give that error.

seadowg replied:
If this error is indeed only on architectures with missing libs (which it looks like and hopefully is), then that pushes us more towards not including Mapbox in all builds (using product flavours or my preferred conditional dependency method). That would allow us to continue to exclude x86 libs as we'd just use the "non mapbox" build types on emulators.

grzesiek2010 replied:
> If this error is indeed only on architectures with missing libs

Yes the crash occurs only on x86 devices (if I remove those libs).

grzesiek2010 commented on collect_app/build.gradle:
```diff
@@ -115,7 +111,7 @@ android {
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.txt'
             resValue("string", "GOOGLE_MAPS_API_KEY", googleMapsApiKey)
-            buildConfigField 'String', "MAPBOX_ACCESS_TOKEN", '"' + mapboxToken + '"'
+            resValue("string", "mapbox_access_token", mapboxAccessToken)
```
This works out of the box. We generate a string named `mapbox_access_token` with a valid access token and mapbox takes it automatically to setup everything that it needs.

seadowg commented:
@grzesiek2010 is the compilation error on CI expected? I'm guessing it's because there isn't a Mapbox API key configured to grab the SDK.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.kt:
```diff
@@ -108,7 +107,6 @@ class ApplicationInitializer(
                 MapView(context).onCreate(null)
             }
             Configuration.getInstance().userAgentValue = userAgentProvider.userAgent
-            MapboxUtils.initMapbox()
```
We just don't need this any longer? Definitely nice if so!

grzesiek2010 replied:
We don not initialize mapbox like that now. That class was also used to handle cases where a user tried to use mapbox on unsupported devices or a user didn't have a valid access token (then null was returned and we were able to handle that cases). Now as all devices are supported because I removed excluding libraries for X86 devices + without a valid downloads token you won't build the app at all. It's still possible to have a supported device + valid downloads token but wrong access token but it doesn't make sense to check such strange cases I think.

seadowg reviewed (dismissed):
Are you able to roll back to Kotlin conversions here? It makes it hard to understand the actual diff. For instance, Git is unable to track the changes you had to make to `MapboxInitializationFragment` https://github.com/getodk/collect/pull/5141/commits/d85866155bd80945245ff9841ceee8f1e50d9486.

We should potentially start doing our Kotlin conversions as follow-up PRs as I'm finding it's making reviews (and reviewing my own changes) much harder.

seadowg commented:
> sharing our tokens so that everyone can build the app including CircleCi
> storing Mapbox 10 in our local repository

I think these feel roughly equivalent: both would work but might result in Mapbox suspending our account/revoking keys, and we'd have to resort to doing something else eventually. An alternative to this that still keeps the build simple would be to require CI (which could be leaked easily unless we stop running CI for PRs) and developer to all have their own Mapbox account/API keys and create their own local `secrets.properties`.

> having a separate flavor

I still think something along these lines is my preferred option. We can add a product flavour for when Mapbox is included and when it isn't and have the former include the real implementation of `MapboxMapFragment` (and other Mapbox classes) and have the latter include a stub one. 

As I've thought about it more however, I think there might be a "simpler" way to do achieve the same goals: move Mapbox code to a module (we want to do this anyway for all the different `MapFragment` implementations) and only include it when we have an access token for the SDK. As an example, in our dependencies for `collect_app` we'd have something like :

```groovy
if (mapboxAccessToken != null) {
    implementation project(':mapbox')
}
```

This would allow us to have a module that would only compile when `mapboxAccessToken` is present. In other words, it's OK if it doesn't compile because of a missing dependency. Then when we're creating the Fragment in `collect_app` we can just load the class dynamically rather than statically:

```java
try {
    return this.getClassLoader().loadClass("org.odk.collect.mapbox.MapboxMapFragment").newInstance();
} catch (ClassNotFoundException e) {
     return null;
} catch (IllegalAccessException e) {
     return null;
} catch (InstantiationException e) {
     return null;
}
```

If we do go down this road (which I'm advocating for), we probably want to look at merging #5139 before this as those changes will be needed to allow `MapboxMapFragment` to be pulled out to a separate module.

grzesiek2010 commented:
> is the compilation error on CI expected? I'm guessing it's because there isn't a Mapbox API key configured to grab the SDK.

exactly there are no `MAPBOX_DOWNLOADS_TOKEN` so the app can't be built

grzesiek2010 commented:
> Are you able to roll back to Kotlin conversions here? It makes it hard to understand the actual diff. For instance, Git is unable to track the changes you had to make to MapboxInitializationFragment https://github.com/getodk/collect/commit/d85866155bd80945245ff9841ceee8f1e50d9486.

I'm not sorry. Generally I always implement changes and once everything is done I add conversion in a separate commit but not this time sorry my mistake. To help you a bit with that class what I changed was:
- I've removed onStart/onResume/onPause/onStop/onSaveInstanceState/onLowMemory/onDestroyView methods because now it's handled automatically https://docs.mapbox.com/android/maps/guides/migrate-to-v10/#simplified-lifecycle-management
- I've updated loading styles https://docs.mapbox.com/android/maps/guides/migrate-to-v10/#map-styles

and that's it.

seadowg commented:
Just a note about something to keep in mind: Chromebooks/Chrome OS devices are pretty likely to have x86. That's a real world example of x86 devices that we didn't really consider the first time we removed Mapbox's x86 libs (and therefore x86 support). I imagine we can get away with not supporting Mapbox on Chromebooks, but we should make sure the release build can start OK on them.

grzesiek2010 committed:
Fixed conflicts

grzesiek2010 committed:
Added new module for mapbox

grzesiek2010 committed:
Moved mapbox code to new module

grzesiek2010 committed:
Updated mapbox to its latest version

grzesiek2010 commented:
@seadowg 
I moved mapbox code to its own module.
I wanted to try load classes dynamically like you said above: https://github.com/getodk/collect/pull/5141#issuecomment-1127891995

The problem is that we use that new module in more places than only to create `MapboxMapFragment` like in your example.
I'm going to list those places so that it's clear:
- creating `MapboxMapFragment` in https://github.com/getodk/collect/pull/5141/files#diff-f424798f5410bd1996038d93031423c03974cb72ff81112f1a58be8a03c228f0R21 we should be ale to handle this in the way you presented above
- creating `MapboxMapConfigurator` in https://github.com/getodk/collect/pull/5141/files#diff-3a00df4aebd5600fae250c3807e97417704f10649504cc5f3cad20652bfc2a3fR61 the same as above
- using MapboxMapConfigurator.DEFAULT_MAP_STYLE in [Defaults](https://github.com/getodk/collect/pull/5141/files#diff-210f5b7fc183e9c0fae301ac06622f84d636afb6d2740b21fa5c4599beeb64a0R58) I can replace it with a raw string `mapbox://styles/mapbox/streets-v11` so that `mapbox` module is not used there 
- `MapBoxInitializationFragment` in [main_menu](https://github.com/getodk/collect/pull/5141/files#diff-4beabbc156b71a2271e4b0d6f53adeb9f490488ab884ca0ac56c2933c5d06547R29) this would need to load that fragment dynamically so that if the `mapbox` module is not included there are no errors. BTW I need to double check if we still need this trick.
- places where we use classes from the new module to setup dagger:
https://github.com/getodk/collect/pull/5141/files#diff-19e8ed4f040ab190989099eb37cd00c559b474b8e802b01260614bf9b628773bR12 https://github.com/getodk/collect/pull/5141/files#diff-a2f0be8cb9ce6618f33c478f6ae0c69fc5e6eefa5807833ffc3f1f324369c5b3R281 this might be a problem

what do you think?

seadowg commented:
> creating MapboxMapConfigurator in https://github.com/getodk/collect/pull/5141/files#diff-3a00df4aebd5600fae250c3807e97417704f10649504cc5f3cad20652bfc2a3fR61 the same as above

Yeah, I think we can just dynamically add the Mapbox source option if the configurator is present.

> using MapboxMapConfigurator.DEFAULT_MAP_STYLE in [Defaults](https://github.com/getodk/collect/pull/5141/files#diff-210f5b7fc183e9c0fae301ac06622f84d636afb6d2740b21fa5c4599beeb64a0R58) I can replace it with a raw string mapbox://styles/mapbox/streets-v11 so that mapbox module is not used there

👍

> MapBoxInitializationFragment in [main_menu](https://github.com/getodk/collect/pull/5141/files#diff-4beabbc156b71a2271e4b0d6f53adeb9f490488ab884ca0ac56c2933c5d06547R29) this would need to load that fragment dynamically so that if the mapbox module is not included there are no errors. BTW I need to double check if we still need this trick.

Yes that sounds right. You'll have to add the Fragment with a transaction rather than loading it in XML using `name`.

> places where we use classes from the new module to setup dagger:
https://github.com/getodk/collect/pull/5141/files#diff-19e8ed4f040ab190989099eb37cd00c559b474b8e802b01260614bf9b628773bR12 https://github.com/getodk/collect/pull/5141/files#diff-a2f0be8cb9ce6618f33c478f6ae0c69fc5e6eefa5807833ffc3f1f324369c5b3R281 this might be a problem

Ooooft yeah that's a problem. I think what we'll have to do there is give the dependency module a constructor with args, and then grab the constructor dynamically. So the module would look like this:

```kotlin
@Module
class MapboxDependencyModule(
    private val settingsProvider: SettingsProvider,
    private val networkStateProvider: NetworkStateProvider,
    private val referenceLayerRepository: ReferenceLayerRepository
) {
    @Provides
    fun providesSettingsProvider(): SettingsProvider {
        return settingsProvider
    }

    @Provides
    fun providesNetworkStateProvider(): NetworkStateProvider {
        return networkStateProvider
    }

    @Provides
    fun providesReferenceLayerRepository(): ReferenceLayerRepository {
        return referenceLayerRepository
    }
}
```

I'd been thinking that giving these modules constructors would be nice anyway - it prevents us from having to create `Collect...` subclasses, so it's just less code overall. You can construct that without a static reference like this in Java:

```java
this.getClassLoader().loadClass("org.odk.collect.mapbox.MapboxDependencyModule").getConstructors()[0].newInstance(
    settingsProvider,
    networkStateProvider,
    referenceLayerRepository
);
```

Given we're using reflection to load and construct classes in multiple places, it might be a good idea to have a helper object that hides all this work. I could imagine a `MapboxClassInstanceCreator` that has methods like `createDependencyModule` and `createMapboxInitializationFragment`.

grzesiek2010 committed:
Use reflection to create MapboxMapFragment

grzesiek2010 committed:
Do not use mapbox module in Defaults

grzesiek2010 committed:
Use reflection to create MapboxMapConfigurator

grzesiek2010 committed:
Use reflection to create MapBoxInitializationFragment

grzesiek2010 committed:
Improved MapboxClassInstanceCreator

grzesiek2010 commented:
I've fixed all steps mentioned above apart from DI. As you know i will be off next two weeks so if this blocks you somehow feel free to take over.

What's left:
- [x] fixing DI
- [x] removing x86 libs
- [x] making sure that everything works together

seadowg committed:
Allow app to build without mapbox token

seadowg committed:
Ignore UnusedResources check on strings only used my Mapbox

seadowg committed:
Remove mapbox from test modules list

seadowg committed:
Only use presence of download token to include/disclude Mapbox

seadowg committed:
Provide dependencies to Mapbox using a basic 'service lookup via shared singleton' mini framework

seadowg committed:
Exclude x86 libs from APKs

This will cause a crash at launch on x86 devices

seadowg committed:
Manually initialize Mapbox to prevent launch crashes on x86

There will still be a crash if Mapbox is selected and then a
map is opened.

seadowg committed:
Don't offer Mapbox on x86 devices

seadowg committed:
Don't offer Mapbox in builds that don't include it

seadowg committed:
Fix non OdkViews not being set on SwipeHandler

seadowg committed:
Try to avoid flakiness when clicking dialog

seadowg commented on .circleci/test_modules.txt:
```diff
@@ -12,5 +12,4 @@ upgrade:testDebug
 permissions:testDebug
 settings:testDebug
 maps:testDebug
-mapbox:testDebug
```
We'll need to be careful to run tests ourselves when making changes to `mapbox`

seadowg commented on README.md:
```diff
@@ -153,15 +153,15 @@ Certain functions in ODK Collect depend on cloud services that require API keys
 
 **Google Maps API**: When the "Google Maps SDK" option is selected in the "User interface" settings, ODK Collect uses the Google Maps API for displaying maps in the geospatial widgets (GeoPoint, GeoTrace, and GeoShape).  To enable this API:
   1. [Get a Google Maps API key](https://developers.google.com/maps/documentation/android-api/signup).  Note that this requires a credit card number, though the card will not be charged immediately; some free API usage is permitted.  You should carefully read the terms before providing a credit card number.
-  1. Edit or create `collect_app/secrets.properties` and set the `GOOGLE_MAPS_API_KEY` property to your API key.  You should end up with a line that looks like this:
```
There's no need `secrets.properties` to be in `collect_app`. Seems more intuitive to me for it to live in the root.

lognaturel replied:
It will be annoying to have to remember to move it and I guarantee some people who maintain forks (maybe even me!) will be super confused and have a hard time finding this note. Does it really seem worth it? I don’t feel strongly but do want to make sure you’ve considered that this has some costs for some types of users.

seadowg replied:
Yeah that's a good point. The main reason for moving it was that `secrets.gradle` needs to access it, and it felt more natural for it to live in the top instead of having to reach down into a project. I'll think of a way to flag it during build (so it's obvious that `secrets.properties` needs to move) or move it back.

seadowg commented on shared/src/main/java/org/odk/collect/shared/injection/ObjectProvider.kt:
```diff
@@ -0,0 +1,35 @@
+ * [ObjectProviderHost.getMultiClassProvider].
+ */
+
+interface ObjectProvider {
```
This is a 35 line mini DI framework. I think there are definitely better solutions than this but let's finish the discussion in #4940 before thinking about that.

lognaturel replied:
In the mean time, consider adding to the comment to state it was introduced to solve a specific problem with Mapbox and is not intended to be used elsewhere.

seadowg replied:
I'm still thinking that if the result of #4940 is "keep Dagger" than I'll just replace this with a shared module for the Dagger setup. I'll leave this right now until we've made some decisions there.

seadowg commented on collect_app/src/main/java/org/odk/collect/android/listeners/SwipeHandler.java:
```diff
@@ -96,10 +97,10 @@ public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float ve
                 // Looks for user swipes. If the user has swiped, move to the appropriate screen.
 
                 // For all screens a swipe is left/right of at least .25" and up/down of less than .25" OR left/right of > .5"
-                int xpixellimit = (int) (ScreenUtils.xdpi(odkView.getContext()) * .25);
```
A few tests were failing here due to `odkView` being null when we swipe back from the end screen. I'm not sure how we've got away with `SwipeHandler` not dealing with `EmptyView` and `FormEndView` for so long.

seadowg commented on collect_app/src/androidTest/java/org/odk/collect/android/support/pages/ListPreferenceDialog.java:
```diff
@@ -17,7 +17,6 @@ public ListPreferenceDialog<T> assertOnPage() {
     public T clickOption(int option) {
-        clickOnString(option);
-        return page.assertOnPage();
+        return clickOnButtonInDialog(option, page);
```
There's an issue referenced in this method.

seadowg commented on mapbox/src/main/AndroidManifest.xml:
```diff
@@ -1,5 +1,25 @@
+            android:exported="false"
+            tools:node="merge">
+
+            <meta-data
```
I couldn't find these initializer classes referenced in the Mapbox docs. This meant I had to debug inside Android library code to work out what initializers were being run. @lognaturel are we able to send some feedback to Mapbox about this?

lognaturel reviewed (dismissed):
I’m delighted to say that now that y’all have done the work, it doesn’t look as scary or complicated as I had feared. I thought I’d want to review this interactively but I’m comfortable with the understanding I got async.

seadowg committed:
Make build explode when secrets.properties is in the wrong place

seadowg committed:
Update README for new Mapbox setup

seadowg commented on README.md:
````diff
@@ -153,18 +153,21 @@ Certain functions in ODK Collect depend on cloud services that require API keys
     ```
 
 **Mapbox Maps SDK for Android**: When the "Mapbox SDK" option is selected in the "User interface" settings, ODK Collect uses the Mapbox SDK for displaying maps in the geospatial widgets (GeoPoint, GeoTrace, and GeoShape).  To enable this API:
   1. [Create a Mapbox account](https://www.mapbox.com/signup/).  Note that signing up with the "Pay-As-You-Go" plan does not require a credit card.  Mapbox provides free API usage up to the monthly thresholds documented at [https://www.mapbox.com/pricing](https://www.mapbox.com/pricing).  If your usage exceeds these thresholds, you will receive e-mail with instructions on how to add a credit card ...
````
It looks to me like creating Mapbox accounts with `@gmail.com` addresses requires a credit card for an ID check (but my own domains seem to work fine) and this blocked @kkrawczyk123 from being able to set up an account. I'm hoping `@soldevelo` emails will work fine, but if not we might want to create an ODK dev account @lognaturel (probably using `@getodk.org`). For now, I've sent @kkrawczyk123 a `secrets.properties` file that with tokens generated in my dev account.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/application/MapboxClassInstanceCreator.kt:
```diff
@@ -0,0 +1,44 @@
+
+    @JvmStatic
+    fun isMapboxAvailable(): Boolean {
+        return try {
```
This could just return `createMapboxMapFragment() != null`

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/geo/MapConfiguratorProvider.java:
```diff
@@ -153,6 +155,16 @@ public static int[] getLabelIds() {
+        );
+
+        return MapboxClassInstanceCreator.isMapboxAvailable()
+                && stream(Build.SUPPORTED_ABIS).anyMatch(abi -> mapboxAbis.contains(abi));
```
You can use method reference here `anyMatch(mapboxAbis::contains)`

grzesiek2010 reviewed (commented):
I've reviewed changes that were added when I was on leave. It looks great only two comments but nothing important.

seadowg committed:
Clean up code detecting Mapbox availability

dbemke commented:
I've found a crash after following those steps:

1. Select MapBox as a source of maps
2. Open form geojson with 1000 points 
[geojson.xml.txt](https://github.com/getodk/collect/files/9002380/geojson.xml.txt)
[thousand-points.geojson.txt](https://github.com/getodk/collect/files/9010719/thousand-points.geojson.txt)


3. Click on Select place in select from geojson
4. While map is loading hit back button

Stack trace from Logcat:
`2022-06-28 15:57:39.761 32216-6979/? E/collect.androi: No implementation found for void com.mapbox.maps.assets.AssetManagerProvider.finalize() (tried Java_com_mapbox_maps_assets_AssetManagerProvider_finalize and Java_com_mapbox_maps_assets_AssetManagerProvider_finalize__)
2022-06-28 15:57:39.761 32216-6979/? E/System: Uncaught exception thrown by finalizer
2022-06-28 15:57:39.761 32216-6979/? E/System: java.lang.UnsatisfiedLinkError: No implementation found for void com.mapbox.maps.assets.AssetManagerProvider.finalize() (tried Java_com_mapbox_maps_assets_AssetManagerProvider_finalize and Java_com_mapbox_maps_assets_AssetManagerProvider_finalize__)
        at com.mapbox.maps.assets.AssetManagerProvider.finalize(Native Method)
        at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:291)
        at java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:278)
        at java.lang.Daemons$Daemon.run(Daemons.java:139)
        at java.lang.Thread.run(Thread.java:923)
2022-06-28 15:57:39.762 32216-6979/? E/collect.androi: No implementation found for void com.mapbox.maps.assets.AssetManagerProvider.finalize() (tried Java_com_mapbox_maps_assets_AssetManagerProvider_finalize and Java_com_mapbox_maps_assets_AssetManagerProvider_finalize__)
2022-06-28 15:57:39.762 32216-6979/? E/System: Uncaught exception thrown by finalizer
2022-06-28 15:57:39.763 32216-6979/? E/System: java.lang.UnsatisfiedLinkError: No implementation found for void com.mapbox.maps.assets.AssetManagerProvider.finalize() (tried Java_com_mapbox_maps_assets_AssetManagerProvider_finalize and Java_com_mapbox_maps_assets_AssetManagerProvider_finalize__)
        at com.mapbox.maps.assets.AssetManagerProvider.finalize(Native Method)
        at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:291)
        at java.lang.Daemons$FinalizerDaemon.runInternal(Daemons.java:278)
        at java.lang.Daemons$Daemon.run(Daemons.java:139)
        at `java.lang.Thread.run(Thread.java:923)``

On Google Maps there is no crash.

srujner commented:
Bug with Submissions: 
- Issue is only reproducing on Mapbox

**Steps to Reproduce:**

1. Go to Project Settings -> Maps and Select Mapbox as Source;
2. Download all Widgets form with Geopoint widget;
3. Fill few submissions and add answer to Geopoint;
4. Have at least two submissions with different statuses;
5. Click on the map icon next to the All Widgets form;
6. Click on the first created Submission and then try to click on another one (Only first submission will be clickable)
![Screenshot_2022-06-28-15-59-00-815_org odk collect android](https://user-images.githubusercontent.com/85225519/176384780-2e9cc505-8f6a-43be-90f1-5817917bb443.jpg)


**Expected Behavior:**
User should be able to click on all submission points

**Additional information:**
- Issue is only reproducing on Mapbox

grzesiek2010 committed:
Deleted unused files after rebasing the branch onto master

grzesiek2010 committed:
Fixed imports

grzesiek2010 committed:
Removed redundant qualifier names

grzesiek2010 committed:
Fixed clicking markers

grzesiek2010 committed:
Fixed imports in BlankFormListMenuDelegateTest

dbemke commented:
Layers in Geoshape and Geotrace widget on Mapbox don’t change the view of the map ( while the map is opened)

**Problem description:**

When layers are applied while the map is opened on Geoshape or Geotrace widget the view of the map doesn’t change. After that clicking back button and next clicking Start Geoshape/Geotrace- the layer is applied. Applying layer Countries closes the form and the user is moved to main menu. 

**Steps to reproduce:**

1. Go to Project Settings -> Maps and Select Mapbox as Source; 
2. Download all Widgets form with Geo widgets; 
3. Go to Geotrace/Geoshape widget
4. Click Start GoeTrace/Geoshape
5. Click on layer button → change layer to e.g. Control Room

**Expected behaviour:**
While maps are opened on Geoshape/ Geotrace widget , changing layers changes the view of the map.

https://user-images.githubusercontent.com/56479916/176446388-44d38a0d-f1ab-478a-a4a1-b29de97da2b9.mp4


**Other information:**
Goepoint widget works well.

grzesiek2010 commented:
@srujner the bug you have reported should be fixed.

@dbemke I can't reproduce the crash you have described. Could you try again with the newest changes?

grzesiek2010 committed:
Fixed creating polylines

grzesiek2010 commented:
@dbemke I've fixed the problem with changing layers that was a tricky one. Good catch!

dbemke commented:
Mapbox with applied vector map layer in Geopoint, Geoshape, Geotrace closes the form.

**Problem description:**

When Mapbox is used as a source of maps and there is a vector layer added and applied, the form crashes. User is moved to main menu.

**Steps to reproduce:**

1. Choose Mapbox as as source of maps in settings
2. Download all Widgets form with Geo widgets
3. Go to Geopoint/Geotrace/ Geoshape widget
4. Click Start
5. Click on layer button
6. Change layer to vector map e.g. 
[countries.mbtiles.txt](https://github.com/getodk/collect/files/9018438/countries.mbtiles.txt)


**Expected behavior:**

Layer is applied. The view of the map is changed. User is able to continue filling in a form.

@grzesiek2010 is already taking a look at this

grzesiek2010 committed:
Fixed vector tiles

grzesiek2010 commented:
@dbemke 

The problem with vector tiles should be fixed.

grzesiek2010 commented:
@srujner @dbemke 

I think this pull request also fixes this issue https://github.com/getodk/collect/issues/3883 Please verify. If not just ignore it wasn't the main goal.

dbemke commented:
@grzesiek2010 
verified on Android 11 and 8. Also verified https://github.com/getodk/collect/issues/3883 . Works as expected.

grzesiek2010 committed:
Fixed early mapbox inizialization

dbemke commented:
Mapbox with offline layers applied when ODK Collect was build without Internet access show a black background instead of a view of a map

**Problem description:**

When ODK Collect is built without access to the Internet and layers are added in directory, Mapbox is used as a source of maps on Geoshape and Geotrace widget or Geopoint with maps appearance,
after choosing a layer there is a black background instead of a view of a map.

**Steps to reproduce:**

    1. Build ODK Collect without access to the Internet
    2. Turn on Internet to download all Widgets form 
    3. Turn off the Internet
    4. Add offline layers in layers directory
    5. Go to ODK Collect app
    6. Choose Mapbox as as source of maps in settings 
    7. Go to All widgets form
    8. Go to Geotrace/ Geoshape widget or Geopoint with maps appearance
    9. Click Start 
    10. Click on layer button 
    11. Choose offline layer
    
**Expected behavior:**

Layer is applied. The view of the map is properly changed.

grzesiek2010 commented:
@dbemke 
The issue with using mapbox offline is now fixed.

srujner commented:
**Problem description:**
Issue can be reproduced only on **Android 5.1** and only for **Mapbox**.

When Mapbox is used as a source of maps and there is a submission created and the user clicks on submission and rotates the screen, the map is zooming out to the maximum value (map of the whole world is visible)

**Steps to reproduce:**

1. Choose Mapbox as a source of maps in settings;
2. Download all Widgets form with Geo widgets;
3. Create at least one submission point;
4. Click on the map icon next to the All Widgets form;
5. Click on the created Submission point and rotate the screen;
6. The map will zoom out to the maximum value

**Expected behavior:**
Map should stay at the same zoom value after rotating the screen

srujner commented:
**Problem description:**
Issue can be reproduced on **all devices** and only for **Mapbox**.

When Mapbox is used as a source of maps and there is a submission created and the user clicks on submission and rotates the screen, the dialog with information about submission disappears.

**Steps to reproduce:**

1. Choose Mapbox as a source of maps in settings;
2. Download all Widgets form with Geo widgets;
3. Create at least one submission point;
4. Click on the map icon next to the All Widgets form;
5. Click on the created Submission point and rotate the screen;
6. The dialog with additional information about submission will disappear.

https://user-images.githubusercontent.com/85225519/176883807-080365c2-3c8c-404f-aaac-7181efb4e2be.mp4

**Expected behavior:**
The dialog should not disappear after rotating the screen

seadowg commented:
It feels like there's only a couple of minor problems left (assuming everything other than the last two issues posted by @srujner are fixed). I'm thinking we can go ahead and merge this to unblock other PRs and create issues for the remaining problems. Sound good @grzesiek2010? If so, @srujner could you create issues for your last couple of comments?

grzesiek2010 commented:
I'm not sure I would rather give the QA team some time till Monday at least if there is nothing big then we can file separate issues for everything.

lognaturel commented:
Maybe y’all can check in when @seadowg starts his Monday? QA might have made good progress by then and it’d be great to unblock other map-related things in flight.

grzesiek2010 committed:
Fixed selecting items on a map

grzesiek2010 committed:
Fixed zooming to points

seadowg commented:
@grzesiek2010 just dismissed reviews so that I remember to have another look at changes. Just request a review when you're done.

grzesiek2010 committed:
Fixed initializing mapbox

srujner commented:
**Tested with Success!**
Verified on Androids: 5.1, 10, 12

Verified cases: 
- Issue [#3883](https://github.com/getodk/collect/issues/3883) is no longer reproducing;
- Five bugs will be created as separate issues, all minor things;
- Regression checks for GeoPoint, GeoTrace, GeoShape widgets;                    
- Regression checks on external geo widgets;
- Regression checks on Forms: geojson, gps_accuracy_threshold, map internal-select, placement_map, TestingGeoPointWidgetWithValue, OSM_buildings;    
- Select Map forms;    
- Offline Layers; 
- Placement by tapping, manual mode and automatic mode
- Google Maps, Mapbox, OSM;         
- Collecting Location in background;             
- View Submissions on map; 
- Sending Submission to Central;
- Switching between maps Google/OSM/Mapbox
- Map view with displayed points    
- Screen rotation, 
- Minimize app, using backward button, 
- Light and Dark mode,

dbemke commented:
**Tested with Success!**
Verified on Androids: 8.1, 11

seadowg reviewed (changes requested):
It looks like b2cea654a7e338398ddd6ecc0f22a12362b21139 and 34089a4eaa54e8b4dab9cccf3a383b85adb58ced make changes to `SelectionMapFragment`, so I'm guessing the issues they address (which I think are https://github.com/getodk/collect/pull/5141#issuecomment-1172217970 and https://github.com/getodk/collect/pull/5141#issuecomment-1172234222 respectively) are not actually Mapbox specific, and I'd suspect are also existing issues rather than new ones caused by this PR (given that the code being changed isn't Mapbox code). Is that right @grzesiek2010?

If so, I think we should leave those commits out and create issues. Those changes will cause some nasty conflicts with #5164 and the issues being fixed may also have already been fixed by those PRs. Additionally, #5164 introduces a way to test interactions with the `BottomSheetBehavior` which would be needed to write a test for b2cea654a7e338398ddd6ecc0f22a12362b21139 which is currently has no coverage.

seadowg commented on mapbox/src/main/java/org/odk/collect/mapbox/MapBoxInitializationFragment.kt:
```diff
@@ -66,13 +66,17 @@ class MapBoxInitializationFragment : Fragment() {
-                AppInitializer.getInstance(context)
-                    .initializeComponent(MapboxMapsInitializer::class.java)
-                it.set(KEY_MAPBOX_INITIALIZED, true)
+        try {
```
I can't work out why this didn't need a `try-catch` when it was called in `MapboxMapFragment`?

grzesiek2010 replied:
It wasn't needed because on a device that does not support mapbox choosing mapbox is not possible so that class `MapboxMapFragment` was never used. However `MapBoxInitializationFragment` was always used if there were keys provided (what casued crashes on emulators for example). I improved it in https://github.com/getodk/collect/pull/5141/commits/24af4510bc9f0c43419c47deb8503001ddb2a7ef and now `try-catch` is not needed indeed because we check both keys and supported abis.

grzesiek2010 commented:
https://github.com/getodk/collect/commit/34089a4eaa54e8b4dab9cccf3a383b85adb58ced yes it was an existing issue. the problem was that if we received current location before loading items on the map then other events like zooming to those points were ignored. However it wasn't visible on master at least on my device because somehow the process was a little bit faster (I think the new map fragment loads slower).

https://github.com/getodk/collect/commit/b2cea654a7e338398ddd6ecc0f22a12362b21139 here again the code was broken but it worked well somehow and only now the bug is visible

So both issues existed but both of them were let's say hidden.

seadowg commented:
> https://github.com/getodk/collect/commit/34089a4eaa54e8b4dab9cccf3a383b85adb58ced yes it was an existing issue. the problem was that if we received current location before loading items on the map then other events like zooming to those points were ignored. However it wasn't visible on master at least on my device because somehow the process was a little bit faster (I think the new map fragment loads slower).
>
> https://github.com/getodk/collect/commit/b2cea654a7e338398ddd6ecc0f22a12362b21139 here again the code was broken but it worked well somehow and only now the bug is visible
>
> So both issues existed but both of them were let's say hidden.

Thanks for the clarification. That makes sense. Can we pull those commits out for now (to a branch so that we don't lose the solutions you came up with), and then we can create issues or @getodk/testers can verify as part of a pass on #5164? That PR reworks `SelectionMapFragment` and adds a lot of tests to `SelectionMapFragment` so a rebase on top of these commits is going to be much harder than fixing the issues (and being able to write tests for them) there instead.

grzesiek2010 commented:
> Can we pull those commits out for now

Those are pretty serious issues to me so I wouldn't release v2022.3 without it but if you want to just exclude it to avoid conflicts and then fix soon - I'm ok with it and I can do that.

grzesiek2010 committed:
Improved checking if mapbox is avaialble

grzesiek2010 committed:
Revert "Fixed selecting items on a map"

This reverts commit b2cea654a7e338398ddd6ecc0f22a12362b21139.

grzesiek2010 committed:
Revert "Fixed zooming to points"

This reverts commit 34089a4eaa54e8b4dab9cccf3a383b85adb58ced.

grzesiek2010 commented:
@seadowg ok I reverted those two commits and I'm going to file separate issues for them.

seadowg commented:
> Those are pretty serious issues to me so I wouldn't release v2022.3 without it but if you want to just exclude it to avoid conflicts and then fix soon - I'm ok with it and I can do that.

Thanks! Agreed, we need to fix before releasing.
